### PR TITLE
weapon_category for martial arts cross-mod compatibility

### LINKED ISSUE #45612: Martials_like: A solution to cross-mod martial arts compatibility
# Is your feature request related to a problem? Please describe.
The +1/+2 weapons from Magiclysm only work with vanilla martial arts, when Mythical Martial Arts would pair so well with them(also applies to Aftershock and C.R.I.T. Expansion, but those weapons aren't magic like some of the MythMAs are).

# Describe the solution you'd like
Ability to add martials_like to a weapon's definition, that when applied adds it to every martial art that the weapon it is referencing also works with.

# Describe alternatives you've considered
Using ``copy-from`` to accomplish it.

# Additional context
```
{
    "type": "GENERIC",
    "id": "mace_plus_one",
    "copy-from": "mace",
    "looks_like": "mace",
    "martials_like": "mace",
    "name": "mace +1",
    "proportional": { "price": 3.0, "bashing": 1.1, "cutting": 1.1, "weight": 0.9 },
    "relative": { "to_hit": 1 }
  }
```
-Adds it to Desert Wind, which has the three vanilla maces as options.

### PR DESCRIPTION
**Pull request by Saicchi**
#### Summary
Infrastructure "weapon_category for martial arts cross-mod compatibility"

#### Purpose of change
Closes #45612

Lets weapons from one mod use martial arts in another mod as long as they have any matching categories.

#### Describe the solution
Add a new field for martial arts
```json
"weapon_category": [ "CATA", "CATB", "CATC" ]
```
and new identical field for weapons
```json
"weapon_category": [ "CATA", "CATJ" ]
```

If **any** of the weapon categories in the weapon are also present in the martial art weapon category, then the weapon is valid for that martial art.

#### Describe alternatives you've considered
Instead of any, weapon is only valid if **all** categories are present.

Replacing `weapon` with `weapon_category`. This would take the flexibily of martial arts made for only a subset of weapons.

#### Testing
No issues found. A weapon from Magiclysm was able to use a martial art from Mythical Martial Arts

#### Additional context
More discussion for this can be found in #48905.

I've also added a few tests to check if a weapon can be used with the martial art.

### CODE REVIEW

**Comment by GuardianDll** [comment]
So it means instead of using `techniques` in each weapon we can just make different categories of weapon martial arts (like "spear", "polearm",  "short blade" etc) and use them? Sound really cool for me

**Comment by Saicchi** [comment]
No. It's just the category the weapon is in, which if matches the martial art can be used.

File: doc/JSON_INFO.md [hunk 1]
```diff
@@ -2304,6 +2304,7 @@ See also VEHICLE_JSON.md
 "price": 100,                                // Used when bartering with NPCs. For stackable items (ammo, comestibles) this is the price for stack_size charges. Can use string "cent" "USD" or "kUSD".
 "price_postapoc": "1 USD",                       // Same as price but represent value post cataclysm. Can use string "cent" "USD" or "kUSD".
 "material": ["COTTON"],                      // Material types, can be as many as you want.  See materials.json for possible options
+"weapon_category": [ "WEAPON_CAT1" ],        // (Optional) Weapon categories this item is in for martial arts.
```
Comment (jbytheway) [inline doc/JSON_INFO.md:2307]: Given that `weapons` is plural, I think it makes sense to also use the plural `weapon_categories` here (but I realize other things like `material` are singular, so we're not entirely consistent).

File: src/martialarts.cpp [hunk 2]
```diff
@@ -848,7 +849,11 @@ bool martialart::has_technique( const Character &u, const matec_id &tec_id ) con
 
 bool martialart::has_weapon( const itype_id &itt ) const
 {
-    return weapons.count( itt ) > 0;
+    return weapons.count( itt ) > 0 ||
+           std::any_of( itt->weapon_category.begin(), itt->weapon_category.end(),
+    [&]( const std::string & weap ) {
+        return weapon_category.count( weap ) > 0;
+    } );
```
Comment (jbytheway) [inline src/martialarts.cpp:856]: This isn't the most efficient way to test for two sets intersecting, but a faster solution would be more verbose, which is probably not worthwhile for something non-performance-critical.

I *might* be inclined to separate this into a new `do_sets_intersect` function.  Makes me wonder if we already have similar code elsewhere.

**Review by jbytheway — COMMENTED** [review]
LGTM.  Couple of minor niggles.

**Comment by cainiaowu** [comment]
Isn't this concept working like "tool quality", an intermediate layer for item group? 
Maybe you can expand tool quality to cover this?

File: src/martialarts.cpp (hunk 2 above)
Comment (Saicchi) [reply src/martialarts.cpp:856]: Having one would definitely be useful. I find myself checking if any member of a `set` is inside another `set` quite a lot.

**Comment by Saicchi** [comment]
I actually forgot why I turned this into a draft...

**Pull request merged**